### DISCUSSION TITLE
Unable to add private nuget feed, as it doesn't specify a RegistrationsBaseUrl

### DISCUSSION THREAD

**Discussion by pingu2k4** [renovatebot/renovate#27830] [Request Help] [answered]
### What would you like help with?

Other

### How are you running Renovate?

Mend Renovate hosted app on github.com

### Please tell us more about your question or problem

Hey,
So I have a private nuget feed, which I am using bytesafe for.
The relevant parts of my config are:

```json
{
  "nuget": {
    "registryUrls": [
      "https://api.nuget.org/v3/index.json",
      "https://[REPLACED].bytesafe.dev/nuget/[REPLACED]/index.json"
    ]
  },
  "hostRules": [
    {
      "hostType": "nuget",
      "matchHost": "https://[REPLACED].bytesafe.dev/nuget/[REPLACED]/index.json",
      "username": "[REPLACED]",
      "encrypted": {
        "password": "[REPLACED - Encrypted using the tool for doing so]"
      }
    }
  ],
}
```

I get an error that a package, hosted within this private repo, was not found (Package lookup failures)
The relevant log messages are underneath, and show that there is no @type of `RegistrationsBaseUrl`, only `RegistrationsBaseUrl/Versioned`.

When visiting the nuget index myself, whilst authenticated, I get the same response also.

Is this nuget feed simply incompatible and something I cannot use? Or is there some additional config I might be able to set to get around this obstacle?

Thanks!

### Logs (if relevant)

<details><summary>Logs</summary>

```

DEBUG: no RegistrationsBaseUrl services found
{
  "url": "https://[REPLACED].bytesafe.dev/nuget/[REPLACED]/index.json"
  "servicesIndexRaw": {
    "version": "3.0.0",
    "resources": [
      {
        "@id": "https://[REPLACED].bytesafe.dev/nuget/[REPLACED]/registrations",
        "@type": "RegistrationsBaseUrl/Versioned",
        "comment": "Get package metadata."
      },
      {
        "@id": "https://[REPLACED].bytesafe.dev/nuget/[REPLACED]/pull",
        "@type": "PackageBaseAddress/Versioned",
        "comment": "Get package content (.nupkg)."
      },
      {
        "@id": "https://[REPLACED].bytesafe.dev/nuget/[REPLACED]/push",
        "@type": "PackagePublish/Versioned",
        "comment": "Push and delete (or unlist) packages."
      },
      {
        "@id": "https://[REPLACED].bytesafe.dev/nuget/[REPLACED]/search",
        "@type": "SearchQueryService",
        "comment": "Filter and search for packages by keyword"
      },
      {
        "@id": "https://[REPLACED].bytesafe.dev/nuget/[REPLACED]/search-3.0.0-beta",
        "@type": "SearchQueryService/3.0.0-beta",
        "comment": "Filter and search for packages by keyword"
      },
      {
        "@id": "https://[REPLACED].bytesafe.dev/nuget/[REPLACED]/search-3.0.0-rc",
        "@type": "SearchQueryService/3.0.0-rc",
        "comment": "Filter and search for packages by keyword"
      }
    ]
  }
}


```

</details>

**Comment by viceice** [comment]
can't be used yet. please provide the `index.json`, so we can maybe find a workaround.

**Reply by pingu2k4** [reply]
The index.json would require authentication. The entire contents of the index.json is output in the log output that I gave in original post - there isn't any more than that sorry.

**Reply by viceice** [reply]
ok, it seems `RegistrationsBaseUrl/Versioned` is invalid[^1] and we disard it. Can you ask your platform provider to set a proper version[^2]? It should be `3.0.0-beta`, `3.0.0-rc` or any `3.0.0`+
[^1]: https://learn.microsoft.com/en-us/nuget/api/service-index#versioning
[^2]: https://learn.microsoft.com/en-us/nuget/api/registration-base-url-resource#versioning

**Reply by pingu2k4** [reply]
Thanks. I've had them add an entry between my previous response and your most recent reply. It now looks like this (with the unchanged parts not pasted, just the addition and the section that follows it)

```
{
  "@id": "https://[REPLACED].bytesafe.dev/nuget/[REPLACED]/registrations",
  "@type": "RegistrationsBaseUrl",
  "comment": "Get package metadata."
},
{
  "@id": "https://[REPLACED].bytesafe.dev/nuget/[REPLACED]/registrations",
  "@type": "RegistrationsBaseUrl/Versioned",
  "comment": "Get package metadata."
},
```

I can't test it yet however, as renovate on runs since this update was performed seems to be retrieving a cached version of the index, which doesn't have this addition. Is there a way to force it to ignore the cache for this temporarily?

**Reply by viceice** [reply]
not on the hosted app, sorry.

**Reply by pingu2k4** [reply]
I bypassed the cache for now by adding an attribute to the end, which has no effect on the response. The change they made for me didn't help, I get the same log only with the additional resource listed. There's something different I would need to ask them to do?

<details><summary>Logs</summary>

```
DEBUG: no RegistrationsBaseUrl services found
{
  "url": "https://[REPLACED].bytesafe.dev/nuget/[REPLACED]/index.json?x=1"
  "servicesIndexRaw": {
    "version": "3.0.0",
    "resources": [
      {
        "@id": "https://[REPLACED].bytesafe.dev/nuget/[REPLACED]/registrations",
        "@type": "RegistrationsBaseUrl",
        "comment": "Get package metadata."
      },
      {
        "@id": "https://[REPLACED].bytesafe.dev/nuget/[REPLACED]/registrations",
        "@type": "RegistrationsBaseUrl/Versioned",
        "comment": "Get package metadata."
      },
      {
        "@id": "https://[REPLACED].bytesafe.dev/nuget/[REPLACED]/pull",
        "@type": "PackageBaseAddress/Versioned",
        "comment": "Get package content (.nupkg)."
      },
      {
        "@id": "https://[REPLACED].bytesafe.dev/nuget/[REPLACED]/push",
        "@type": "PackagePublish/Versioned",
        "comment": "Push and delete (or unlist) packages."
      },
      {
        "@id": "https://[REPLACED].bytesafe.dev/nuget/[REPLACED]/search",
        "@type": "SearchQueryService",
        "comment": "Filter and search for packages by keyword"
      },
      {
        "@id": "https://[REPLACED].bytesafe.dev/nuget/[REPLACED]/search-3.0.0-beta",
        "@type": "SearchQueryService/3.0.0-beta",
        "comment": "Filter and search for packages by keyword"
      },
      {
        "@id": "https://[REPLACED].bytesafe.dev/nuget/[REPLACED]/search-3.0.0-rc",
        "@type": "SearchQueryService/3.0.0-rc",
        "comment": "Filter and search for packages by keyword"
      }
    ]
  }
}
```

</details>

**Reply by viceice** [reply]
we need `RegistrationsBaseUrl/<version>` with values already mentioned. The version must be a valid semver v2.


https://github.com/renovatebot/renovate/blob/b8ffc95a7a83c8c56dc7952079b62ac2d31ba863/lib/modules/datasource/nuget/v3.ts#L64

**Reply by pingu2k4** [reply]
Hey. I asked them to add what it looked like was needed and it works, however only up until the next point at which there is a new issue. The authentication I am providing is correct, as it wouldn't be able to load the index without it (and I have triple checked that it has all the access it should have, even generated a new token and provided that).

The problem now is that it loads the index, sees the URL to search on, attempts to use that but appears to be getting 401's. I cannot replicate that my end however, the access details I provided, if it gets through to the index should also be all it needs to not get a 401 on the search page. Here are my log outputs:

<details><summary>Logs</summary>

```
DEBUG: GET https://[REPLACED].bytesafe.dev/nuget/[REPLACED]/registrations/[REPLACED].dataaccessextensions/index.json = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=401 retryCount=0, duration=486)
DEBUG: Datasource unauthorized
{
  "datasource": "nuget"
  "packageName": "[REPLACED].DataAccessExtensions"
  "url": "https://[REPLACED].bytesafe.dev/nuget/[REPLACED]/registrations/[REPLACED].dataaccessextensions/index.json"
}

DEBUG: Failed to look up nuget package [REPLACED].DataAccessExtensions
```

</details>

**Reply by viceice** [reply]
what's your host rule?

**Reply by pingu2k4** [reply]
<details><summary>Config</summary>

```
  "nuget": {
    "registryUrls": [
      "https://api.nuget.org/v3/index.json",
      "https://[REPLACED].bytesafe.dev/nuget/[REPLACED]/index.json"
    ]
  },
  "hostRules": [
    {
      "hostType": "nuget",
      "matchHost": "https://[REPLACED].bytesafe.dev/nuget/[REPLACED]/index.json",
      "username": "[Username]",
      "encrypted": {
        "password": "Encrypted password using https://app.renovatebot.com/encrypt"
      }
    }
  ],
```

</details>

**Reply by viceice** [reply, marked as answer]
remove the `index.json` from `matchHost` so that hostrule matches all urls starting with `https://[REPLACED].bytesafe.dev/nuget/[REPLACED]/`

**Reply by pingu2k4** [reply]
Ahh, this worked. Thank you so much! :)
